core/domain/platform_parameter_domain.py

```diff
@@ -696,35 +711,31 @@ def _validate_feature_flag(self):
         """
         if self._data_type != 'bool':
             raise utils.ValidationError(
-                'Data type of feature flags must be bool, got %s '
+                'Data type of feature flags must be bool, got \'%s\' '
                 'instead.' % self._data_type)
-        if self._metadata.stage not in ALLOWED_FEATURE_STAGES:
+        if self._metadata.feature_stage not in ALLOWED_FEATURE_STAGES:
             raise utils.ValidationError(
-                'Invalid feature stage, expected on of %s, got %s.' % (
-                    ALLOWED_FEATURE_STAGES, self._data_type))
+                'Invalid feature stage, got \'%s\', expected one of %s.' % (
+                    self._metadata.feature_stage, ALLOWED_FEATURE_STAGES))
+
         enabling_rules = [
             rule for rule in self._rules if rule.value_when_matched]
         for rule in enabling_rules:
-            if not rule.has_server_mode_filter():
-                raise utils.ValidationError(
-                    'Rules that enable a feature must have a server_mode '
-                    'filter.')
-            mode_filters = [
-                mode_filter for mode_filter in rule.filters
-                if mode_filter.type == 'server_mode']
-            for mode_filter in mode_filters:
-                value_list = (
-                    mode_filter.value if isinstance(mode_filter.value, list)
-                    else [mode_filter.value])
-                if self._metadata.stage == FEATURE_STAGES.dev:
+            server_mode_filters = [
+                server_mode_filter for server_mode_filter in rule.filters
+                if server_mode_filter.type == 'server_mode']
+            for server_mode_filter in server_mode_filters:
+                server_modes = [
+                    value for _, value in server_mode_filter.conditions]
```

Comment:
```suggestion
                server_modes = [
                    value for value in server_mode_filter.conditions.values()]
```

Slightly more pythonic version.

Reply:
`.conditions` is list of tuples, I think `.values()` is only applicable for dicts.

Reply:
Ah, I missed that. Thanks.